qmi/core/transport.py

```diff
@@ -612,155 +635,314 @@ def _safe_socket(self) -> socket.socket:
-                raise QMI_EndOfInputException(
-                    "Reached end of input from socket {}".format(format_address_and_port(self._address)))
+                b, addr = self._read_from_socket(nbytes - nbuf)
+
+            except (BlockingIOError, socket.timeout) as err:
+                raise QMI_TimeoutException(
+                    f"Timeout after {nbuf} bytes while expecting {nbytes}"
+                ) from err
             self._read_buffer.extend(b)
+            nbuf = len(self._read_buffer)
+            if timeout is not None:
+                tremain = tstart + timeout - time.monotonic()
+                if tremain < 0:
+                    raise QMI_TimeoutException(f"Timeout after {nbuf} bytes while expecting {nbytes}")
 
-    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
-        try:
-            ret = self.read(nbytes, timeout)
-        except QMI_TimeoutException:
-            ret = bytes(self._read_buffer)
-            self._read_buffer = bytearray()
-        except QMI_EndOfInputException:
-            if not self._read_buffer:
-                raise
-            ret = bytes(self._read_buffer)
-            self._read_buffer = bytearray()
+        ret = bytes(self._read_buffer[:nbytes])
+        self._read_buffer = self._read_buffer[nbytes:]
         return ret
 
-    def discard_read(self) -> None:
+    def read_until(self, message_terminator: bytes, timeout: Optional[float] = None) -> bytes:
+        ret = super().read_until(message_terminator, timeout)
+        if len(ret):
+            return ret
+
         self._check_is_open()
-        self._read_buffer = bytearray()
-        self._safe_socket.settimeout(0)
+        self._safe_socket.settimeout(timeout)
+        tstart = time.monotonic()
         while True:
             try:
-                b = self._safe_socket.recv(4096)
-            except socket.timeout:
-                # no more bytes available
-                break
-            except BlockingIOError:
-                # no more bytes available
-                break
-            if not b:
-                # end of stream
-                break
-
-
-def _is_valid_hostname(hostname: str) -> bool:
-    """Return True if the specified host name has valid syntax."""
-    # source: https://stackoverflow.com/questions/2532053/validate-a-hostname-string
-    if len(hostname) > 255:
-        return False
-    if hostname[-1] == ".":
-        hostname = hostname[:-1]  # strip exactly one dot from the right, if present
-    parts = hostname.split(".")
-    if re.match(r"^[0-9]+$", parts[-1]):
-        return False  # numeric TLD, not a domain name
-    allowed = re.compile(r"(?!-)[A-Z0-9-]{1,63}(?<!-)$", re.IGNORECASE)
-    return all(allowed.match(x) for x in parts)
+                # Read from socket. NOTE: Reading up to 512 bytes here for TCP,
+                # as larger packet size causes issues with Tektronix 5014 AWG.
+                b, addr = self._read_from_socket(512)
```

Comment:
major: except for the 512. The rest seems exactly the same. Isn't there a nice way to push this method to the base class. maybe have a MAX_PACKET_SIZE or something constant that changes based on concrete transport.

Reply:
This method could be unified into base class like that. But I would like a solution that works also for the `read` method as well (using it in that one, too). I would push this on another ticket, like the "Rename QMI_Transport read call names and modify their inputs for more logical naming", or make a new ticket altogether.

Reply:
Similar to above, making MAX_PACKET_SIZE would work:
```python
class SocketTransport
  def read_untill(...)
    ...
    b, addr = self._read_from_socket(self.MAX_READ_PACKET_LEN )
    ...

class UdpTransport(SocketTransport):
    MAX_READ_PACKET_LEN = 4096

class TcpTransport(SocketTransport):
    MAX_READ_PACKET_LEN = 512
```

Reply:
Do you think it's a good idea to limit our transports based on certain instruments? What if certain instrument can work with larger packet sizes in TCP transport? I feel like it would be better if that packet size could be injected by the instrument itself if it knows it has some limitation.

But like you said, let's make an extra ticket for that case.

Reply:
Well I think it is necessary for this case if we have an (old) instrument that can handle less than the others. I think 1024 would be better as it is lower than the regular MTU of 1500 and a power of 2 number, but indeed didn't work with the Tektronix. I'll fix this now with the `MAX_PACKET_SIZE = 512` and the suggested (also by @rbudhrani) packet size injection from the instrument will be handled in another ticket. Then for some instruments we could also allow jumbo frames etc.